enhancements/cluster-scope-secret-volumes/csi-driver-host-injections.md

```diff
@@ -118,53 +98,23 @@ the credentials, I want to stop updating content within the existing pods consum
-as part of an OLM based install at the time of this writing is immature, at least for some reviewers.
-- However, other reviewers wonder if we are in fact close from having either OLM or CVO operators being able to require
-an OLM operator to be present.
-- Using OLM avoids further increasing the install payload size.
-- Using OLM avoids unexpected hiccups in the feature's installation impacting initial cluster install or cluster upgrade.
-
-#### CSI starting points
-
-**WE ARE GOING WITH EPHEMERAL MODE**
+For the developer preview release, the CSI driver will be installable via OLM.
+This release channel is meant to be a preview only option and allow the CSI driver to be installed on OpenShift outside of the normal release cycle.
 
-**NO SecretProvideClass IS NEEDED**
+When the CSI driver is ready for general availability, it will be installed on OpenShift via the cluster storage operator and an associated delivery operator - see the [CSI driver install proposal](/enhancements/storage/csi-driver-install.md).
+The delivery operator and OLM operator can be one and the same, with appropriate modifications depending on the context.
 
-With those simplifying decisions, a CSI driver with those characteristics still meets our needs.  And the storage
-SMEs recommend that approach to us.
-
-Crafting the CSI driver in this way also helps us deal with SELinux.  Specifically (thanks to @jsafrane for details):
+#### CSI starting points
 
-- when CRI-O starts any container, it updates the SELinux labels on all volumes for that container, where those labels are unique to the container
-- in the context of k8s, multiple containers within a pod have the same filesystem access
-- so, only that specific pod can read files on that volume
-- however, a file can only have one SELinux label
-- so two or more pods cannot access the same file
-- but if each pod has its own file, its own copy, of the shared data, things work easily from a SELinux perspective
-- this is where using ephemeral CSI storage comes in
-- using ephemeral based storage allows for use of a tmpfs (virtual memory filesystem) for storage
-- tmpfs is fast and has negligible overhead
-- it will have to reconstructed if the Pod moves
-- but how it eases dealing with SELinux protection is the bonus we are taking
-- so each pod gets its own tmpfs
-- the kubelet leverages tmpfs in a similar fashion already
+The projected resource CSI driver will provision ephemeral volumes and take advantage of upstream's ephemeral CSI storage mechanisms.
+Ephemeral storage allows the CSI driver to use tmpfs for storage while providing appropriate protection via SELinux labels.
+Each pod will have its own copy of the shared resource that the CSI driver mounts in, and there will be a 1:1 correspondence between the pod and the SELinux label applied to the mounted files.
+The process will be similar to what exists in core Kubernetes when Secrets and ConfigMaps are mounted into a pod.
```

Comment:
In Kubernetes we have some tricks to make sure content of the volume is changed in an atomic way, so the Pod gets always correct content when a secret is updated in the middle of Pod's `read()`. Applying something like that here would be beneficial. See https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util/atomic_writer.go

(btw, as consequence, we made Secrets and ConfigMap volumes as read-only in the pods, because users could fool the atomic writer to access the host filesystem by rogue symlinks).

Reply:
Yeah I looked at the atomic_writer.go when originally crafting things. but I did not use it explicitly because of the k8s magic linking stuff, where I did not want to follow that path, IIRC.  And then at the time I did not copy/paste/simplify that logic per se moving forward.

But in hindsight yeah I know see what you are getting at.  And agree we should revisit.

There is also the read only stuff.  At the moment, the driver does not support read only, but between this discussion and other discussion in https://github.com/openshift/enhancements/pull/733 I think that needs to be added to the productization epic.

So @adambkaplan if you agree, I think the new questions/features to note in the EP and track are 
- do we want to update after creation at all
- if so, opt into updating or opt out of updating
- approximate / reuse as much as possible k8s/k8s atomic_writer.go.  Though using it as is means vendoring in https://github.com/kubernetes/kubernetes which at first blush I do not want to take on, so approximate/copy seems the thing.

Reply:
With minor refactoring it could be possible to move the atomic writer to `kubernetes/utils`, which has almost no deps and is easily consumable. If we can satisfy "_Used by multiple Kubernetes repositories_" criteria.

Reply:
> So I read this as:
> 
>     1. If we only sync content at volume creation, said volume can be writeable. We should support read-only as an option.
> 

Not 100% sure I'm parsing your item 1)  correctly @adambkaplan but I *think* you are on the right path.  I'm going to tweak your wording a bit, to better reconcile with what the current code does, and how code changes might manifest.

Today, since we always send Secret/ConfigMap updates from the controller watch to the tmpfs mount that backs the CSI volume, we always create that mount with read-write permissions.

However, if we introduce a mode where we never send Secret/ConfigMap updates after the initial provisioning of the CSI volume and its underlying tmpfs mount, we could set things up where the tmpfs mount that baacks the CSI volume is read only.

Implementation detail: I'll have to refresh my memory when I try to implement this, but IIRC, we will need to create a writeable tmpfs that we initially seed with the in memory ConfigMap / Secret contents from the controller, then bind mount that to a second read only tmpfs mount, where that second read only tmpfs mount is what is provided to the Pod for its CSI volume.  We cannot just create a read only tmpfs mount, then write data to it.

>     2. If we do sync content, we should ensure the volume is read-only.
> 

Less sure about this statement @adambkaplan 

So I'm assuming here you are separating the fact that he CSI Pod volume is marked read-only from whether the tmpfs mount that backs the CSI volume is read-write, because if it was read-only, we cannot write the new ConfigMap/Secret data as part of "syncing" content.

If you were separating the read write nature of the pod volume from the read write nature of the underlying tmpfs mount, I'm not sure it works that way.  By that, I mean I'm not sure if there is a mechanism other than OS level file permissions that prevents logic running in the Pod from writing to the volume.

If there is such a mechanism I'm unaware of it.

Or as an alternative, perhaps it is possible to do "the sync"  by a) unmount the existing tmpfs volume with the old ConfigMap/Secret contents (where it is a bind mount of another tmpfs volume we can update ... a "side care writeable tmpfs mount"), b) update our "side car writeable tmpfs mount", then c) bind mount a new read only tmpfs from the "side care writeable tmpfs" with the new ConfigMap/Secret contents to the same path as the tmpfs we unmounted in a)

^^ seems doable assuming it is not a gross abuse of tmpfs and it does not violate the k8s volume contract in some what I am unaware of 

If you were referring to such a method @adambkaplan, then "OK", though it took some reading between the lines to get there.

> 
> @gabemontero @jsafrane is this correct?

Reply:
Unmounting existing tmpfs *is* gross. There is some time where an application does not see any config / secret file and it may break the app in unexpected ways.

You should ensure that whatever you mount inside NodePublish is read-only. You can mount a read-write tmpfs *somewhere*, prepare the projected files there, and then bind-mount it read-only to the directory given to you in `NodePublish`. Thinking about it, it becomes hard to maintain, because you need keep track of these read-write mounts and they must be able to survive CSI driver pod stop + restart (e.g. during CSI driver update).

Reply:
Yeah @jsafrane  I was worried about the grossness of the unmount / mount to get updates.  Also thanks for the confirm on the 2 tmpfs / bind mount 1 to the other for read only notion I mentioned earlier in this thread.

I do not disagree on the maintenance wrt driver restart, but I have in fact dove into the CSI driver pod stop / restart scenario, persisting state related to live pods, etc.  I do think this is a doable progression.

To bubble back up though @adambkaplan , IMO your original item 2) needs to be changed.

If we do sync content, the volume needs to be read-write.

As we do not want to go down the unmount / remount path.

But that does mean we need some form of k8s atomic writer equivalent.

Which circles back to considering turning off the sync content feature for now, as builds does not need it, and then drill down on the need of it for the other scenarios like cert rotation, etc. which we solve later, with more runway.

Reply:
OK so I've been able to experiment with things in this space a bit and have some finding to reports, including some corrections to things I stated in this threads earlier.

- the 2 tmpfs with the "top level" bind mount to the pod destination approach does work
- by writing to the "lower tmpfs", we can change content at that layer, but still have pod destination be read only
- I even added some `touch /data/foo` bash invocations in my pod to confirm that failed when the pod volume was marked read only, but it succeeded when the pod volume was read write
- my prior statements on the direct correlation between tmp fs read/write perms and the pods volume read/write perms was incorrect because I forgot that the kubelet add its own mount on top of things (`/var/kubelet/....` or some such)

So moving forward, we can definitely support a mode where our driver refreshes the content (to support cert rotation scenarios for example), but we do *NOT* allow pod volumes based on our driver to be provisioned unless the `readOnly: true`.

I was able to validate both elements today with pre-merge branch I have.

I'll next move onto updating / validating the e2e's I have around changing data, restarting the driver / pod, etc..  But at first blush, I'm reasonably confident we can maintain this additional element of state, and bypass the need for a k8s/k8s atomic writer equivalent.

Reply:
One amendment to ^^

We don't have to worry about multiple writers, but IIRC there is still the concern of a Pod trying to *read* its mounted content while we are updating it.

Reply:
So my restart the driver tests have uncovered that more than I hoped for will be involved with adding the second tmpfs layer, but still being able to change the data for the Pod after the driver restart.

I'm still experimenting but I suspect the "gross" unmount/remount of tmpfs options previously discussed above might be needed.

I'll report back when I land on something definitive.

But falling back on the previously discussed staging plan where we support the build scenario with full read only and push out support for rotating certs, etc. which require update may be where we land short term.